sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -349,15 +326,14 @@ public void validate(PipelineOptions options) {
           getWindowedWrites(),
           "Must use windowed writes when applying %s to an unbounded PCollection",
           WriteFiles.class.getSimpleName());
-      // The reason for this is https://issues.apache.org/jira/browse/BEAM-1438
-      // and similar behavior in other runners. Runners can choose to ignore this check and perform
-      // runner determined sharding for unbounded data by overriding the option
-      // `withRunnerDeterminedShardingUnboundedInternal`.
-      if (!getWithRunnerDeterminedShardingUnbounded()) {
+      // Sharding used to be required due to https://issues.apache.org/jira/browse/BEAM-1438 and
+      // similar behavior in other runners. Some runners may support runner determined sharding now.
+      // Check merging window here due to https://issues.apache.org/jira/browse/BEAM-12040.
+      if (input.getWindowingStrategy().needsMerge()) {
```

Comment:
Won't this be a behavior regression for runners that do not support auto-sharding ? Current behavior is a clean failure at job submission after this it will be a failure at runtime ?

Reply:
@kennknowles is there a better way to do this without regressions and without runner specific overrides ?

How about adding overrides to runners that do not support auto-sharding to detect this condition and fail early ?

Reply:
There won't be a regression in the sense that runner determined sharding was not supported for unbounded data before (so no performance comparison to make). All existing pipelines should have shard explicitly set and will continue to use the implementation for fixed sharding. New pipelines may observe different behavior if they omit the shard parameter (which will not get an error).

Reply:
(By "this condition", I mean unbounded writes without num_shards or a shards provider)

Reply:
I think it's a regression since failing early (for new pipelines) is much more desired over failing at runtime.

Reply:
It would not fail at runtime. Runner determined sharding is achieved via GroupIntoBatches for which Dataflow does special optimizations. For runners who do not optimize it, the default sharding strategy in GroupIntoBatches will be used, which is one shard per thread:
https://github.com/apache/beam/blob/a8cd05932bed9b2480316fb8518409636cb2733b/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java#L179

Reply:
It would not fail at runtime. Runner determined sharding is achieved via GroupIntoBatches for which Dataflow does special optimizations. For runners who do not optimize it, the default sharding strategy in GroupIntoBatches will be used, which is one shard per thread:
https://github.com/apache/beam/blob/a8cd05932bed9b2480316fb8518409636cb2733b/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java#L179

Reply:
If it is the concern about the potential poor performance of the default sharding strategy is concerning, is there a better way to signal that?

Reply:
I see. I think replacing a definite failure condition with a working but inefficient path should be OK.